lib/pos_mempool_persister.go

```diff
@@ -0,0 +1,205 @@
+package lib
+
+import (
+	"bytes"
+	"github.com/deso-protocol/core/storage"
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+	"sync"
+	"time"
+)
+
+const (
+	DbMempoolContextId = "transactions"
+	eventQueueSize     = 10000
+)
+
+type MempoolEventType int
+
+const (
+	MempoolEventAdd MempoolEventType = iota
+	MempoolEventRemove
+	MempoolEventExit
+)
+
+type MempoolPersisterStatus int
+
+const (
+	MempoolPersisterStatusRunning MempoolPersisterStatus = iota
+	MempoolPersisterStatusNotRunning
+)
+
+// MempoolPersister is responsible for persisting transactions in the mempool to the database. Whenever a transaction is
+// added or removed from the mempool, the MempoolPersister is notified via the OnMempoolEvent callback. The Persister
+// will then add the event to a queue. Periodically, the transaction queue is flushed to the database and all the cached
+// transactions are persisted.
+type MempoolPersister struct {
+	sync.Mutex
+	stopGroup  sync.WaitGroup
+	startGroup sync.WaitGroup
+
+	frequencyMilliseconds int
+	status                MempoolPersisterStatus
+	dbCtx                 *storage.DatabaseContext
+	eventQueue            chan *MempoolEvent
+	updateBatch           []*MempoolEvent
+}
+
+func NewMempoolPersister(dbCtx *storage.DatabaseContext, frequencyMilliseconds int) *MempoolPersister {
+	return &MempoolPersister{
+		frequencyMilliseconds: frequencyMilliseconds,
+		status:                MempoolPersisterStatusNotRunning,
+		dbCtx:                 dbCtx,
+		eventQueue:            make(chan *MempoolEvent, eventQueueSize),
+	}
+}
+
+func (mp *MempoolPersister) Start() {
+	if mp.status == MempoolPersisterStatusRunning {
+		return
+	}
+
+	mp.reset()
+	mp.stopGroup.Add(1)
+	mp.startGroup.Add(1)
+	go mp.run()
+	mp.startGroup.Wait()
+	mp.status = MempoolPersisterStatusRunning
+}
+
+func (mp *MempoolPersister) run() {
+	mp.startGroup.Done()
+	for {
+		select {
+		case event := <-mp.eventQueue:
+			switch event.Type {
+			case MempoolEventAdd, MempoolEventRemove:
+				mp.Lock()
+				mp.updateBatch = append(mp.updateBatch, event)
+				mp.Unlock()
+
+			case MempoolEventExit:
+				mp.status = MempoolPersisterStatusNotRunning
+				mp.stopGroup.Done()
```

Comment:
I think you also want to close the event queue channel here, no? It signals to the caller that the channel won't accept any more events. I know you `reset()` everything later on, but the existing instance of the channel is still open until it's garbage collected
```suggestion
				close(mp.eventQueue)
				mp.stopGroup.Done()
```